### PR TITLE
Add Encrypted Server Name Indication for TLS 1.3

### LINKED ISSUE #2096: Add Encrypted Server Name Indication for TLS 1.3
As this Encrypted Server name is added to more and more browsers I believe it would be a good addition to add the ability to dissect tls packets with the with the client hello containing a [Encrypted SNI](https://datatracker.ietf.org/doc/draft-ietf-tls-esni/).

### PR DESCRIPTION
**Pull request by mrsylerpowers**
Added ability to dissect and use [TLS ESNI](https://datatracker.ietf.org/doc/draft-ietf-tls-esni/)
fixes #2096

### CODE REVIEW

**Comment by guedou** [comment]
flake8 is not happy:
```
scapy/layers/tls/extensions.py:160:33: E128 continuation line under-indented for visual indent
scapy/layers/tls/extensions.py:162:32: E128 continuation line under-indented for visual indent
scapy/layers/tls/extensions.py:162:67: W291 trailing whitespace
scapy/layers/tls/extensions.py:164:32: E128 continuation line under-indented for visual indent
scapy/layers/tls/extensions.py:165:80: E501 line too long (87 > 79 characters)
scapy/layers/tls/extensions.py:167:32: E128 continuation line under-indented for visual indent
scapy/layers/tls/extensions.py:169:32: E128 continuation line under-indented for visual indent
scapy/layers/tls/extensions.py:169:57: E231 missing whitespace after ','
scapy/layers/tls/extensions.py:171:32: E128 continuation line under-indented for visual indent
scapy/layers/tls/extensions.py:205:1: E302 expected 2 blank lines, found 1
scapy/layers/tls/extensions.py:210:80: E501 line too long (81 > 79 characters)
scapy/layers/tls/extensions.py:212:32: E128 continuation line under-indented for visual indent
scapy/layers/tls/extensions.py:214:32: E128 continuation line under-indented for visual indent
scapy/layers/tls/extensions.py:215:80: E501 line too long (87 > 79 characters)
scapy/layers/tls/extensions.py:217:32: E128 continuation line under-indented for visual indent
scapy/layers/tls/extensions.py:219:32: E128 continuation line under-indented for visual indent
scapy/layers/tls/extensions.py:221:32: E128 continuation line under-indented for visual indent
```

**Comment by guedou** [comment]
Also, could you add some unit tests as done here https://github.com/secdev/scapy/blob/master/scapy/contrib/gtp.uts#L57 for disecting and instanciating.

**Comment by mrsylerpowers** [comment]
Hey @guedou where should these test be? In the tls1.3 file or the tls or in the contrib without a python file to go with it unlike all of the other files there?

**Comment by guedou** [comment]
You can add them at the end of this file https://github.com/secdev/scapy/blob/master/test/tls13.uts

Thanks

File: scapy/layers/tls/extensions.py [hunk 1]
```diff
@@ -178,6 +204,27 @@ class TLS_Ext_ServerName(TLS_Ext_PrettyPacketList):                 # RFC 4366
                                    length_from=lambda pkt: pkt.servernameslen)]
 
 
+class TLS_Ext_EncryptedServerName(TLS_Ext_PrettyPacketList):
+    name = "TLS Extension - Encrypted Server Name"
+    fields_desc = [ShortEnumField("type", 0xffce, _tls_ext),
+                   ShortField("len", None),
+                   EnumField("cipher", None, _tls_cipher_suites),
+                   ShortEnumField("key_exchange_group", None,
+                                  _tls_named_groups),
```
Comment (mrsylerpowers) [inline scapy/layers/tls/extensions.py]: Key Exchange Entry Feild here instead of feilds like this?

File: scapy/layers/tls/extensions.py [hunk 2]
```diff
@@ -149,6 +151,30 @@ def guess_payload_class(self, p):
         return Padding
 
 
+class EncryptedServerName(Packet):
+    name = "Encrypted HostName"
+    fields_desc = [ShortEnumField("type", 0xffce, _tls_ext),
+                   ShortField("len", None),
+                   EnumField("cipher", None, _tls_cipher_suites),
+                   ShortEnumField("key_exchange_group", None,
```
Comment (mrsylerpowers) [inline scapy/layers/tls/extensions.py]: Key Exchange Entry Feild here instead of feilds like this?

File: scapy/layers/tls/extensions.py [hunk 3]
```diff
@@ -149,6 +151,30 @@ def guess_payload_class(self, p):
         return Padding
 
 
+class EncryptedServerName(Packet):
```
Comment (guedou) [inline scapy/layers/tls/extensions.py]: Is this needed?

Comment (mrsylerpowers) [reply scapy/layers/tls/extensions.py]: Nope Just following the same format as all of the other extensions. Will remove

**Comment by mrsylerpowers** [comment]
@guedou I believe this should be ready for review

**Comment by gpotter2** [comment]
We're probably going to wait for https://github.com/secdev/scapy/pull/2132 to be merged before merging any TLS-related PRs. Thanks for the PR though !

Scapy only implements TLS 1.3 - draft 18 in its current state. It has diverted a lot from the current RFC, and we probably want to focus into making that work first.

**Comment by gpotter2** [comment]
Hi !
Could you rebase against master ?
`git pull --rebase upstream master`

**Comment by mrsylerpowers** [comment]
@gpotter2 the 'ext' attribute of a TLS object doesn't seem to be present when using the old crypto library and seems to fail to build. Is this a problem I should be concerned about

**Comment by gpotter2** [comment]
Thanks for your PR & sorry for the delay

**Pull request closed without merging**